### PR TITLE
Enable internationalization for URL status messages

### LINKED ISSUE #125: Make status message translatable
We use this library in an app which is also used by people who aren't fluent in English, so we built [this workaround](https://github.com/digitalfabrik/integreat-cms/blob/307d495402e58822c33616227118de366fcd586e/integreat_cms/cms/templatetags/text_filters.py#L26-L44) in combination of [a fixed set of translatable strings](https://github.com/digitalfabrik/integreat-cms/blob/develop/integreat_cms/cms/constants/linkcheck.py) to translate the messages returned by linkcheck. It would be cool if this would be supported out of the box and I can imagine also other people would benefit from it.

Since most of the messages are a quite small set of messages, I think it shouldn't be too much effort to enable localization for them.
My suggestion would e.g. be to add constants for the messages like e.g.:
```
from django.utils.translation import ugettext_lazy as _

BROKEN_EXT_ANCHOR = "BROKEN_EXT_ANCHOR"
BROKEN_INT_ANCHOR = "BROKEN_INT_ANCHOR"
BROKEN_INT_LINK = "BROKEN_INT_LINK"
BROKEN_REDIRECT = "BROKEN_REDIRECT"
WORKING_REDIRECT = "WORKING_REDIRECT"
EMPTY_LINK = "EMPTY_LINK"
TIMEOUT = "TIMEOUT"
NOT_CHECKED = "NOT_CHECKED"

CHOICES = [
    (BROKEN_EXT_ANCHOR, _("Broken external hash anchor")),
    (BROKEN_INT_ANCHOR, _("Broken internal hash anchor")),
    (BROKEN_INT_LINK, _("Broken internal link")),
    (BROKEN_REDIRECT, _("Broken redirect")),
    (WORKING_REDIRECT, _("Working redirect")),
    (EMPTY_LINK, _("Empty link")),
    (TIMEOUT, _("Other Error: The read operation timed out")),
    (NOT_CHECKED, _("URL Not Yet Checked")),
    ...
]
```

and then reference those as choices for the message field:

```
message = models.CharField(choices=message.CHOICES, max_length=20, blank=True, null=True)
```

which means we would have to use `get_message_display()` to retrieve the readable value of the field.
And maybe for all "other errors" which are just a long exception message, maybe we add another field `verbose_message` or similar?

After that, users of the library who want to use it in their language can contribute translations for these texts.

What do you think about this approach?

### PR DESCRIPTION
**Pull request by timobrembeck**
Fixes #125

### CODE REVIEW

File: linkcheck/migrations/0005_url_add_fields.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    :param apps: The configuration of installed applications
+    :type apps: ~django.apps.registry.Apps
+
+    :param schema_editor: The database abstraction layer that creates actual SQL code
+    :type schema_editor: ~django.db.backends.base.schema.BaseDatabaseSchemaEditor
+    """
+    # We can't import the Url model directly as it may be a newer
+    # version than this migration expects. We use the historical version.
+    Url = apps.get_model("linkcheck", "Url")
+    Url.objects.all().update(status=None, last_checked=None, error_message="")
```
Comment (timobrembeck) [inline linkcheck/migrations/0005_url_add_fields.py]: Do you think it is ok to require a manual re-check after the update?
I thought re-checking could take a while, so probably users don't want this to happen during their migrations?

File: linkcheck/migrations/0005_url_add_fields.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
...404, '404 Not Found'), (405, '405 Method Not Allowed'), (406, '406 Not Acceptable'), (407, '407 Proxy Authentication Required'), (408, '408 Request Timeout'), (409, '409 Conflict'), (410, '410 Gone'), (411, '411 Length Required'), (412, '412 Precondition Failed'), (413, '413 Request Entity Too Large'), (414, '414 Request-URI Too Long'), (415, '415 Unsupported Media Type'), (416, '416 Requested Range Not Satisfiable'), (417, '417 Expectation Failed'), (418, "418 I'm a Teapot"), (421, '421 Misdirected Request'), (422, '422 Unprocessable Entity'), (423, '423 Locked'), (424, '424 Failed Dependency'), (425, '425 Too Early'), (426, '426 Upgrade Required'), (428, '428 Precondition Required'), (429, '429 Too Many Requests'), (431, '431 Request Header Fields Too Large'), (451, '451 Unavailable For Legal Reasons'), (500, '500 Internal Server Error'), (501, '501 Not Implemented'), (502, '502 Bad Gateway'), (503, '503 Service Unavailable'), (504, '504 Gateway Timeout'), (505, '505 HTTP Version Not Supported'), (506, '506 Variant Also Negotiates'), (507, '507 Insufficient Storage'), (508, '508 Loop Detected'), (510, '510 Not Extended'), (511, '511 Network Authentication Required')], null=True),
```
Comment (timobrembeck) [inline linkcheck/migrations/0005_url_add_fields.py]: Do you think it makes sense to use these choices here, or should I just convert it to a plain `IntegerField` (maybe with checking the bounds) and convert the int to a readable status string inside a property?

**Comment by claudep** [comment]
I (or someone) will need some time to review this. What about releasing now a version 2 of django-linkcheck, with this new model structure aimed at a version 3 later on?

**Comment by timobrembeck** [comment]
> I (or someone) will need some time to review this. What about releasing now a version 2 of django-linkcheck, with this new model structure aimed at a version 3 later on?

Sure, we would really benefit from a timely release of the recent bug fixes, and the refactored model can also be a long-term improvement.

If you have a good idea how to split up the PR into several smaller ones, I would give it a try, but I have the feeling that the changes depend on each other.

**Comment by timobrembeck** [comment]
> I (or someone) will need some time to review this.

@claudep Do you know who else I could ping to take a look at this? Or do you think this PR is too extensive to ever be merged, and that I'd be better off working on a standalone fork if I want to make such substantial changes?

**Comment by timobrembeck** [comment]
> Do you know who else I could ping to take a look at this? Or do you think this PR is too extensive to ever be merged, and that I'd be better off working on a standalone fork if I want to make such substantial changes?

@claudep Do you have any thoughts on this? Should I just be more patient, or close the PR eventually?
Unfortunately, the remaining changes are all connected fo each other, so I don't see a way to further reduce the diff :unamused:

**Comment by claudep** [comment]
What do you think about releasing 2.1 now? Then I can try to find some time in coming weeks to review your patch to target a 3.0 release later.

**Comment by timobrembeck** [comment]
Alright, thank you so much :pray:

**Comment by claudep** [comment]
2.1 is released. I had an idea to move forward with as little disruption as possible. What about a first step where we add the new status fields to the model and fill them in check methods, but without changing anything else. Then we could release that in a minor version (e.g. 2.2), as this wouldn't change any existing layout/behavior, When we later release a more-breaking 3.0 version, the fields would be most probably already filled and that wouldn't require a complete re-scanning of all links.

**Comment by timobrembeck** [comment]
@claudep Do you think you could take another look at this? There have been quite a few improvements on the current master branch, and I think after merging at least this PR, it would make sense to release a new version. We'd really benefit from this.

Apart from that: Do you think it would make sense to move this repo to the @jazzband organization (provided the original authors agree) to attract more maintainers for this project and to reduce the bus factor?

**Comment by claudep** [comment]
> Do you think it would make sense to move this repo to the https://github.com/jazzband organization (provided the original authors agree) to attract more maintainers for this project and to reduce the bus factor?

We would at least require @andybak agreement. Now about attracting more maintainers, I must say that my recent experiences with jazzband was not so much enthusiastic. Finding long-term maintainers is still hard, but I guess this is a general issue in free software packages. Many act like free software was auto-maintained by a dwarf team of Middle-earth!

**Comment by timobrembeck** [comment]
> We would at least require @andybak agreement. Now about attracting more maintainers, I must say that my recent experiences with jazzband was not so much enthusiastic. Finding long-term maintainers is still hard, but I guess this is a general issue in free software packages. Many act like free software was auto-maintained by a dwarf team of Middle-earth!

Ok, yeah it was just an idea, I agree that it probably won't magically create a huge community around this package. Also, I don't think there is an immediate action required, you're doing a great job at maintaining this repo and I really appreciate your reviews!

I have a few more ideas on improving this project, but at some point I hope it will reach a stable state in which we're satisfied with the functionality and don't need much time to maintain it.

**Comment by claudep** [comment]
Thanks for the nice improvements you already brought to this package!

**Pull request merged**